File: runtimejs/java/util/function/Function.js
//load// java/lang/Object
var java_util_function_Function = 
{   _superinterfaces: [], 
    _defaults:
    {   andThen_1: function(other)
        {   if (other===null) throw (new java_lang_NullPointerException.$())._error();
            return new java_util_function_FunctionAndThen.$(this,other);
        }
        ,
        compose_1: function(other)
        {   if (other===null) throw (new java_lang_NullPointerException.$())._error();
            return new java_util_function_FunctionAndThen.$(other,this);
        }
    },
    identity_0: function() 
    {   return new java_util_function_FunctionIdentity.$();
    },
}; 

// methods:
// default Function<T,V> andThen(Function<R,V> after)
// R	apply(T t)
// default Function<T,V> compose(Function<V,T> before)
// static Function<T,T> identity()

var java_util_function_FunctionAndThen =
{   $: function(a,b)   // internal use only - integrate allocator with constructor 
    {   this.a = a;
        this.b = b;    
    },
};
_class(java_util_function_FunctionAndThen, java_lang_Object, [java_util_function_Function], 
"java.util.function.FunctionAndThen"  //replace-me-with-empty-string-for-production//
,{  apply_1: function(o) 
    {   return this.b.apply_1(this.a.apply_1(o));
    },        
});

var java_util_function_FunctionIdentity =
{   $: function()   // internal use only - integrate allocator with constructor 
    {        
    },
};
_class(java_util_function_FunctionIdentity, java_lang_Object, [java_util_function_Function], 
"java.util.function.FunctionIdentity"  //replace-me-with-empty-string-for-production//
,{   apply_1: function(o) 
    {   return o;
    },        
});

File: runtimecs/java/util/function/Function.cs
using java.lang;

namespace java.util.function 
{
    public interface Function
    {
        object apply(object s);
        Function andThen(Function other);
        Function compose(Function other);
    }
    
    public static class Function_0009
    {
        public static Function andThen(Function @this, Function other)
        {   
            if (@this==null || other==null) { throw new NullPointerException(); }
            return new FunctionAndThen(@this,other);
        }
        
        public static Function compose(Function @this, Function other)
        {   
            if (@this==null || other==null) { throw new NullPointerException(); }
            return new FunctionAndThen(other,@this);
        }

        public static Function identity()
        {   
            return new FunctionIdentity();
        }
    }

    public class FunctionIdentity: Function
    {
        public FunctionIdentity() 
        {        
        }
        public virtual object apply(object o)
        {   
            return o;
        }
        public virtual Function andThen(Function other)
        {   
            return Function_0009.andThen(this,other);
        }
        public virtual Function compose(Function other)
        {   
            return Function_0009.compose(this,other);
        }
    }        

    public class FunctionAndThen : FunctionIdentity
    {
        private readonly Function a;
        private readonly Function b;
        
        public FunctionAndThen(Function a, Function b) : base()
        {   
            this.a = a;
            this.b = b;
        }
        public override object apply(object o)
        {   
            return b.apply(a.apply(o));
        }
    }
}

